[macOS] Rename viewId to viewIdentifier

dkwingsmt opened the pull request:
This PR adds `FlutterViewController` with new API `viewIdentifier` and renames all occasions that mention "viewId" to "viewIdentifier", to align with the requirement for the iOS shell.

A new typedef `FlutterViewIdentifier` is also added. ~~The problem is, we don't have a file to contain this typedef. Currently I put them in a new file called `common.h`. I'm open to other suggestions.~~ It has been moved to `FlutterViewController.h`.
* Another alternative is to not use the type def, but use `int64_t` directly. I'm ok with this choice too, since honestly it's just a 64 int.

Also, macOS's definition for `kFlutterImplicitViewId` has been removed in favor of the one in `common/constants.h`.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

cbracken commented on shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:
```diff
@@ -106,15 +106,16 @@ @interface FlutterEngine () <FlutterBinaryMessenger, FlutterMouseCursorPluginDel
     NSMutableDictionary<NSString*, FlutterEngineRegistrar*>* pluginRegistrars;
 
-- (nullable FlutterViewController*)viewControllerForId:(FlutterViewId)viewId;
+- (nullable FlutterViewController*)viewControllerForId:(FlutterViewId)viewIdentifier;
```
Should we be updating the method name here with `Id` -> `Identifier`?

cbracken commented on shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:
```diff
@@ -106,15 +106,16 @@ @interface FlutterEngine () <FlutterBinaryMessenger, FlutterMouseCursorPluginDel
  */
-- (void)registerViewController:(FlutterViewController*)controller forId:(FlutterViewId)viewId;
+- (void)registerViewController:(FlutterViewController*)controller
+                         forId:(FlutterViewId)viewIdentifier;
```
These should probably be `withId` or `withIdentifier`.

cbracken commented on shell/platform/darwin/macos/framework/Source/FlutterViewController.mm:
```diff
@@ -330,7 +330,7 @@ @implementation FlutterViewController {
 }
 
-@synthesize viewId = _viewId;
+@synthesize viewIdentifier = _viewId;
```
We should probably update the ivar name for consistency.

cbracken commented on shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:
```diff
@@ -110,6 +111,39 @@ FLUTTER_DARWIN_EXPORT
+ * See:
+ * https://api.flutter.dev/flutter/dart-ui/PlatformDispatcher/implicitView.html
+ */
+- (BOOL)isImplicitView;
```
I understand why we need to expose this in dart:ui to the framework, but why do we need this in the public app-facing interface?

dkwingsmt replied:
Removed. Thanks for pointing out.

cbracken commented:
Oh wow the lint bot upset about something that's not formatting :) Looks like the linter caught a couple leaks -- CFRelease the values allocated by CGEventCreateScrollWheelEvent and CGEventCreateCopy.

loic-sharma commented on shell/platform/darwin/common/framework/Headers/common.h:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Unique identifier for views.
+ *
+ * View IDs are generated by the embedder, and are not interpreted by the
+ * embedder API in any way.
```
These details are very specific to the embedder/engine integration.

Is this is a header file public to end users? If yes, I would consider removing this sentence. End users likely don't know what the embedder API is. They likely don't need to know who generates these IDs either.

loic-sharma commented on shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:
```diff
@@ -110,6 +111,39 @@ FLUTTER_DARWIN_EXPORT
+ * The identifier for this view controller, if it is attached.
+ *
+ * The identifier is assigned when the view controller is attached to a
+ * `FlutterEngine. See `FlutterEngine#addViewController:` for details.
```
```suggestion
 * `FlutterEngine`. See `FlutterEngine#addViewController:` for details.
```

dkwingsmt commented:
@cbracken @loic-sharma I should have addressed all comments. I removed `isImplicitView`, and renamed supposedly all places that mention the view ID to be `viewIdentifier` (while carefully avoiding the platform view IDs). Can you take a look again?

cbracken commented on shell/platform/darwin/macos/framework/Source/FlutterViewController.mm:
```diff
@@ -469,9 +467,14 @@ - (void)setBackgroundColor:(NSColor*)color {
+- (BOOL)isImplicitView {
   NSAssert([self attached], @"This view controller is not attached.");
-  return _viewId;
+  return self.viewIdentifier == kFlutterImplicitViewId;
```
Should kFlutterImplicitViewId be renamed to kFlutterImplicitViewIdentifier?

dkwingsmt replied:
It's an internal variable so we can keep it more brief, and I kind of want to keep the variable name aligned with the counterparts in other platforms. Should I change it to Identifier too?

dkwingsmt replied:
Actually, there's a `kFlutterImplicitViewId` defined in `common/constants.h` and I've switched the macOS embedder to use that one. The definition in `FlutterView.h` has been removed.

cbracken commented on shell/platform/darwin/common/framework/Headers/common.h:
```diff
@@ -0,0 +1,15 @@
+#include <objc/objc.h>
+
+/**
+ * A unique numeric value for a view of an engine.
```
Why "of an engine"?

dkwingsmt replied:
I want to express the following two points:
* The IDs are unique within an engine (and are not unique between different engines).
* The views are the engine views, not platform views.

I'm not sure if these points are worth mentioning at all or are there better ways. Do you have a suggestion?

dkwingsmt replied:
That's awesome! Thank you! I'll use this one.

cbracken commented on shell/platform/darwin/common/framework/Headers/common.h:
Is there no existing header file we can add the view to? It would be nice to avoid adding new headers, especially very generic ones. If for some reason there's no way to add to an existing header, we should name this one consistently with the other public headers in FlutterMacOS.framework.

/cc @jmagman @vashworth who I believe may have work in progress to make iOS and macOS framework headers more consistent.

dkwingsmt replied:
I agree, and that's my major question for this PR. Thank you for tagging them.

jmagman replied:
I agree, adding new public headers is a lot of overhead, particularly for one typedef.  I think `FlutterViewController` would be a better place for this.

In the future, if you add a public header, make sure it matches the naming convention of the other headers, and that you add it to the umbrella header[s]:
https://github.com/flutter/engine/blob/main/shell/platform/darwin/macos/framework/Headers/FlutterMacOS.h
https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Headers/Flutter.h

dkwingsmt replied:
Thanks for the suggestion. It was originally because `FlutterPluginRegistrarMacOS.h` also needs it, which doesn't import `FlutterViewController.h`. But let's leave this to a problem for the future. For now, I've moved it to `FlutterViewController.h`.

cbracken commented on shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:
```diff
@@ -868,13 +877,27 @@ - (FlutterCompositor*)createFlutterCompositor {
+  NSCAssert(controller.engine == nil,
+            @"The FlutterViewController is unexpectedly attached to "
+            @"engine %@ before initialization.",
+            controller.engine);
```
Why are we using NSCAssert here instead of NSAssert? NSCAssert should be reserved for plain C functions.

dkwingsmt replied:
Fixed! It was mistake due to the code movement.

cbracken reviewed (approved):
![LGTM stamp from a Japanese personal seal](https://github.com/flutter/engine/assets/351029/19586e93-7d17-487e-9b88-01517268e719)
